File: apps/api/routers/concepts.py
"""Concepts router — the concept-graph read model, the gap list + triage, and the rebuild trigger.

Read-only over the vocabulary by decision (ADR-017 A1): the graph observes, the Manage-keywords
routes (in the library router) edit. The one write here is the rebuild *trigger*, which runs the
same idempotent runner the CLI does (202 + poll, ADR-017 B1). Route order is preserved from the
pre-split module — ``/api/concepts/gaps`` stays declared before ``/api/concepts/{concept_id}/…``.
"""

from __future__ import annotations

import threading
from typing import Any

from fastapi import APIRouter, HTTPException, Request

from apps.api.models import (
    ConceptGraphPayload,
    ConceptPresencePayload,
    GapListItemPayload,
    GapTriageRequest,
)
from apps.api.services import _graph_rebuild_status_dict, _GraphRebuildStatus, log

router = APIRouter()


@router.get("/api/concepts/graph")
def get_concept_graph() -> ConceptGraphPayload:
    """The concept-graph read model: nodes + edges + communities + gaps + a staleness verdict.

    **404 when the skeleton has never been built** — ``skeleton.json`` is a gitignored,
    regenerable sidecar, so a fresh clone legitimately has none. That is an empty state the UI
    answers with the rebuild button below, not a server fault; the detail says so.

    Ids are concept **UUIDs** throughout (nodes, edge endpoints, gap anchors, community
    members); ``label`` rides only on the node. One id space, deliberately — see KI-15.
    """
    from doc_assistant.knowledge.concept_graph_view import load_graph_view

    view = load_graph_view()
    if view is None:
        raise HTTPException(
            status_code=404,
            detail="concept graph not built yet — run a rebuild to create it",
        )
    return ConceptGraphPayload.from_view(view)


@router.get("/api/concepts/gaps")
def get_gap_list() -> list[GapListItemPayload]:
    """The first-class gap list (ROADMAP E5, ADR-004/ADR-017 C1): every detected corpus gap
    with its concept label resolved and its **effective** triage status (a user override wins).

    A pure sidecar read — no model, no LLM. Empty list when no gaps are built (0-doc /
    pre-``build_gaps`` — the normal first-run state), never an error. Presentation order (the
    RG-014 strong-kinds-first lens) is the client's; this serves the detector's order."""
    from doc_assistant.knowledge.concept_graph_view import load_gap_list

    return [GapListItemPayload.from_item(item) for item in load_gap_list()]


@router.post("/api/concepts/gaps/triage")
def triage_gap(body: GapTriageRequest) -> dict[str, bool]:
    """Record (or reset) a user's triage verdict on one gap (ADR-017 C1, E5).

    The verdict lives in the ``gap_triage`` override sidecar keyed on ``(concept_id, kind)``,
    so it survives ``build_gaps``'s delete-and-rebuild of the deterministic rows — a dismissal
    means something across the acquire loop. ``surfaced`` resets to the detector's default
    (deletes the override). Idempotent; no validation beyond the enum (the gap need not still
    exist — triaging a gap the next rebuild will drop is harmless and self-cleaning)."""
    from doc_assistant.knowledge.gaps import set_gap_status

    set_gap_status(body.concept_id, body.kind, body.status)
    return {"ok": True}


@router.get("/api/concepts/{concept_id}/presence")
def get_concept_presence(concept_id: str) -> list[ConceptPresencePayload]:
    """Where one concept appears: its documents + the chunk keys that mention it.

    Per-concept, not bulk — the view renders one neighbourhood at a time (ego-first), and the
    chunk-key set grows with the vocabulary. An unknown concept returns ``[]`` (a concept that
    is present nowhere is indistinguishable here, by design — the vocabulary is the authority
    on existence)."""
    from doc_assistant.knowledge.concept_graph_view import load_concept_presence

    return [ConceptPresencePayload.from_presence(p) for p in load_concept_presence(concept_id)]


@router.post("/api/concepts/graph/rebuild", status_code=202)
def rebuild_concept_graph(request: Request) -> dict[str, Any]:
    """Rebuild the concept skeleton (Node A: ~7s, zero-LLM, deterministic, idempotent).

    202 + poll, mirroring ``POST /api/ingest`` — the repo's established shape for a
    derived-data build triggered from the app (ADR-017 B1). 409 while one is already running.
    """
    app_ = request.app
    status: _GraphRebuildStatus = app_.state.graph_rebuild_status
    with app_.state.graph_rebuild_lock:
        if status.state == "running":
            raise HTTPException(status_code=409, detail="graph rebuild already running")
        status.state = "running"
        status.message = None

    def _worker() -> None:
        try:
            version = app_.state.rebuild_graph_fn()
        except Exception as e:  # a failed rebuild must not kill the thread silently
            log.exception("graph_rebuild_failed")
            with app_.state.graph_rebuild_lock:
                status.state = "error"
                status.message = str(e)
            return
        with app_.state.graph_rebuild_lock:
            status.graph_version = version
            status.message = "graph rebuilt"
            status.state = "done"

    threading.Thread(target=_worker, name="graph-rebuild", daemon=True).start()
    return _graph_rebuild_status_dict(app_)


@router.get("/api/concepts/graph/rebuild/status")
def concept_graph_rebuild_status(request: Request) -> dict[str, Any]:
    return _graph_rebuild_status_dict(request.app)

File: apps/desktop/src/lib/core/api/concepts.ts
// Thin `concepts` client for the desktop API — fetch + parsing only, no business logic.
// Pairs with apps/api/routers/concepts.py and apps/api/models/concepts.py; see
// docs/architecture.md, section "apps/ — the domain spine".

import { API_BASE, errorDetail } from './_base'
import type {
  ConceptGraph,
  ConceptPresence,
  GapListItem,
  GapStatus,
  GraphRebuildStatus,
} from '../types'

// Concept graph (docs/specs/feature-concept-graph.md, ADR-017). Read-only render model + a 202+poll
// rebuild, mirroring the ingest job. A 404 is the NORMAL first run — skeleton.json is gitignored, so
// a fresh clone has none; the caller renders an empty state offering a rebuild, not an error.
export async function getConceptGraph(): Promise<ConceptGraph | null> {
  const r = await fetch(`${API_BASE}/api/concepts/graph`)
  if (r.status === 404) return null
  if (!r.ok) throw new Error(await errorDetail(r, 'concept graph'))
  return (await r.json()) as ConceptGraph
}
/** The first-class gap list (E5, ADR-004): every detected gap with its concept label + effective
 *  triage status. Pure sidecar read, $0. Empty list = no gaps built (0-doc / pre-build), not an error. */
export async function getGapList(): Promise<GapListItem[]> {
  const r = await fetch(`${API_BASE}/api/concepts/gaps`)
  if (!r.ok) throw new Error(await errorDetail(r, 'gap list'))
  return (await r.json()) as GapListItem[]
}
/** Record (or reset) a user's triage verdict on one gap (ADR-017 C1). `surfaced` resets to the
 *  detector's default. Keyed on (concept_id, kind) so it survives the deterministic rebuild. */
export async function triageGap(
  conceptId: string,
  kind: string,
  status: GapStatus,
): Promise<void> {
  const r = await fetch(`${API_BASE}/api/concepts/gaps/triage`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ concept_id: conceptId, kind, status }),
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'triage gap'))
}
/** Where one concept appears, down to the chunk keys — the ego view's navigation payload. Served
 *  per-concept (not bulk in the graph) so one neighbourhood's chunks load at a time. */
export async function getConceptPresence(conceptId: string): Promise<ConceptPresence[]> {
  const r = await fetch(`${API_BASE}/api/concepts/${encodeURIComponent(conceptId)}/presence`)
  if (!r.ok) throw new Error(await errorDetail(r, 'concept presence'))
  return (await r.json()) as ConceptPresence[]
}
/** Trigger a rebuild (202). Deterministic, ~7s, zero-LLM; poll getGraphRebuildStatus until done. */
export async function rebuildConceptGraph(): Promise<GraphRebuildStatus> {
  const r = await fetch(`${API_BASE}/api/concepts/graph/rebuild`, { method: 'POST' })
  if (!r.ok) throw new Error(await errorDetail(r, 'rebuild concept graph'))
  return (await r.json()) as GraphRebuildStatus
}
export async function getGraphRebuildStatus(): Promise<GraphRebuildStatus> {
  const r = await fetch(`${API_BASE}/api/concepts/graph/rebuild/status`)
  if (!r.ok) throw new Error(`graph rebuild status failed: ${r.status}`)
  return (await r.json()) as GraphRebuildStatus
}

// Taxonomy (docs/specs/feature-taxonomy-view.md, ADR-028 increment 2b). The curated field forest +
// coverage read model (2a) and its placement write-path. $0, zero-LLM — deterministic sidecar
// reads/writes. Edits *placement* only (concept→field `in_field` edges + document→field); concept
// vocabulary + field structure stay elsewhere (ADR-017 A1 / ADR-019 D11).
